const DayOff = require('../models/dayoffModel');
const Employee = require('../models/employeeModel'); // voc칡 precisa ter esse model
const { convertToCamelCase } = require('../utils/convertToCamelCase');
const Joi = require('joi');



// Definindo o esquema de valida칞칚o para o DayOff
const dayOffSchema = Joi.object({
  employee_id: Joi.number().integer().required(),
  day: Joi.date().iso().required().messages({
    'data.base': 'O campo "day" deve ser uma data v치lida no formato ISO 8601'
  }),
  reason: Joi.string().max(255).required()
});

exports.getAll = async (req, res) => {
  try {
    const dayoffs = await DayOff.getAll();
    res.json(dayoffs);

  } catch (err) {
    console.error(err);
    res.status(500).json({ error: 'Erro interno no servidor'});
  }
};

exports.create = async (req, res) => {
  console.log(req.body);  // Adicione isso para verificar os dados recebidos
  const { error } = dayOffSchema.validate(req.body);
  if (error) return res.status(400).json({ error: error.details[0].message });

  const { employee_id, day } = req.body;

  try {
    // 游녢 Verifica se o funcion치rio existe
    const employee = await Employee.findById(employee_id);
    if (!employee) {
      return res.status(404).json({ error: 'Funcion치rio n칚o encontrado' });
    }

    // Verifica se j치 existe um "day off" para o mesmo funcion치rio na mesma data
    const existingDayOff = await DayOff.findByEmployeeAndDay(employee_id, day );
    if (existingDayOff) {
      return res.status(400).json({ error: 'O funcion치rio j치 possui um dayoff nesta data' });
    }

    // Cria칞칚o do novo day off
    const newDayOff = await DayOff.create(req.body);
    res.status(201).json(newDayOff);
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: 'Erro interno no servidor' });
  }
};

exports.delete = async (req, res) => {
  try {
    const id = req.params.id;
    const result = await DayOff.delete(id);
    if (!result) {
      return res.status(400).json({error: 'Folga n칚o encontrada' });
    }
    res.status(204).send();
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: 'Erro ao excluir folga'});
  }
}
  
